addons/test_event_full/tests/test_event_mail.py

```diff
@@ -0,0 +1,37 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo.addons.event.tests.common import TestEventCommon
+from odoo.exceptions import ValidationError
+
+class TestTemplateRefModel(TestEventCommon):
+
+    def test_template_ref_model_constraint(self):
+        mail_template = self.env['mail.template'].create({
+            'name': 'test template',
+            'model_id': self.env['ir.model']._get('event.registration').id
+        })
+
+        sms_template = self.env['sms.template'].create({
+            'name': 'test template',
+            'body': 'Body Test',
+            'model_id': self.env['ir.model']._get('event.registration').id
+        })
+
+        with self.assertRaises(ValidationError):
```

Comment:
This could probably be done in a loop and test all combinations, like
```
for notif_type, template_type, valid in [
    ('mail'), ('mail.template'), True),
    ('mail'), ('sms.template'), False),
    ...
]:
    with self.subTest(...):
        # do test
```